llvm/lib/SYCLLowerIR/SYCLConditionalCallOnDevice.cpp

```diff
@@ -0,0 +1,108 @@
+  SmallVector<Function *, 4> FCallers;
+  for (Function &F : M.functions()) {
+    if (F.getName().contains("call_if_on_device_conditionally_helper") ||
+        !F.getName().contains("call_if_on_device_conditionally"))
+      continue;
+
+    FCallers.push_back(&F);
+  }
+
+  int FCallerIndex = 0;
+  for (Function *FCaller : FCallers) {
+    // Find call to @CallableXXX in call_if_on_device_conditionally function
+    // (FAction). FAction should be a literal (i.e. not a pointer)
+    Function *FAction = nullptr;
+    for (BasicBlock &BB : *FCaller) {
+      for (Instruction &I : BB) {
+        if (CallInst *callInst = dyn_cast<CallInst>(&I)) {
+          Value *calledValue = callInst->getCalledOperand();
+          FAction = dyn_cast<Function>(calledValue);
+        }
+      }
+    }
+
+    if (!FAction)
+      continue;
+
+    // Create a new function type with an additional function pointer argument
+    std::vector<Type *> NewParamTypes;
+    Type *FActionType = FAction->getType();
+    NewParamTypes.push_back(
+        PointerType::getUnqual(FActionType)); // Add function pointer to FAction
+    FunctionType *OldFCallerType = FCaller->getFunctionType();
+    for (Type *Ty : OldFCallerType->params()) {
+      NewParamTypes.push_back(Ty);
+    }
+
+    FunctionType *NewFCallerType =
+        FunctionType::get(OldFCallerType->getReturnType(), NewParamTypes,
+                          OldFCallerType->isVarArg());
+
+    // Create a new function with the updated type and rename it to
+    // call_if_on_device_conditionally_GUID_N
+    Twine NewFCallerName = Twine(FCaller->getName()) + "_" + UniquePrefix +
+                           "_" + Twine(FCallerIndex);
+    // Also change to external linkage
+    Function *NewFCaller = Function::Create(
+        NewFCallerType, Function::ExternalLinkage, NewFCallerName, &M);
+
+    NewFCaller->setCallingConv(FCaller->getCallingConv());
+
+    // Replace all calls to the old function with the new one
+    for (auto &U : FCaller->uses()) {
+      if (auto *Call = dyn_cast<CallInst>(U.getUser())) {
+        std::vector<Value *> Args;
+        Args.push_back(
+            FAction); // Add the function pointer as the first argument
+        for (unsigned i = 0; i < Call->arg_size(); ++i) {
+          Args.push_back(Call->getArgOperand(i));
+        }
+
+        // Create the new call instruction
+        CallInst *NewCall =
+            CallInst::Create(NewFCaller, Args, /*	NameStr = */ "", Call);
+        NewCall->setCallingConv(Call->getCallingConv());
+        NewCall->setDebugLoc(Call->getDebugLoc());
+
+        // Replace the old call with the new call
+        Call->replaceAllUsesWith(NewCall);
+        Call->eraseFromParent();
+      }
+    }
+
+    // Remove the body of the new function
+    NewFCaller->deleteBody();
+
+    // Remove the old function from the module
+    FCaller->eraseFromParent();
+
+    FCallerIndex++;
+  }
+
+  return PreservedAnalyses::all();
```

Comment:
I doubt that it is correct, considering that we have changed call graph. I would say `::none()` is a safer option here.